### ISSUE TITLE
Extraction ouput  of a deep layer/without applying activation 

### ISSUE THREAD

**Issue by sarkarachintya** [ageron/handson-ml2#99]
Hi Aurelien,

Your book is nice. After reading your book, I trained a DNN using Keras defined as.
Now I would like to extract the output from a hidden layer for a given "input" without applying the "activation function". Could you please provide sample code for that. It will really help me.
 Best,
Achintya

```python
model = Sequential()
model.add(Dense(1024, input_dim=627, activation='sigmoid'))
model.add(Dense(1024, activation='sigmoid'))
model.add(Dense(1024, activation='sigmoid'))
model.add(Dense(1024, activation='sigmoid'))
model.add(Dense(1024, activation='sigmoid'))
model.add(Dense(1024, activation='sigmoid'))
model.add(Dense(1024, activation='sigmoid'))
model.add(Dense(n_outputs, activation='sigmoid'))
sgd = keras.optimizers.SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', optimizer=sgd, metrics=['accuracy'])

model.fit(X_train, y_train, epochs=3, batch_size=100)
```

**Comment by ageron** [comment]
Hi @sarkarachintya ,

Thanks for your question, and sorry for the late response.
The [sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function) is [invertible](https://en.wikipedia.org/wiki/Inverse_function), meaning that there exists a function which "reverts" it. Indeed, assuming z = sigmoid(x) = 1 / (1 + exp(-x)), you can see that:

* 1/z = 1 + exp(-x)
* so exp(-x) = 1/z - 1
* so log(exp(-x)) = log(1/z - 1)
* so -x = log(1/z - 1)
* so x = -log(1/z - 1)
* so x = -log((1 - z) / z)
* so x = log(z / (1 - z))

This function is called the [logit](https://en.wikipedia.org/wiki/Logit).

So all you need to do is to compute the output of any layer you want, then apply the logit function to the result, and that will give you what you want.

Now to get the result of any layer, you can build a model contain all layers up to that layer, like this:

```python
def logit(y):
    return np.log(y / (1 - y))

first_5_layers = keras.models.Sequential(model.layers[:5])
y_probas = first_5_layers.predict(X_new)
y_logits = logit(y_probas)
```

However, not all activation functions are invertible, for example ReLU is not invertible when x <= 0. So if you want a solution that works with all activation functions, one option is to have separate activation layers:

```python
model = Sequential()
model.add(Dense(1024, input_dim=627))
model.add(keras.layers.Activation("relu"))
model.add(Dense(1024))
model.add(keras.layers.Activation("relu"))
model.add(Dense(1024))
model.add(keras.layers.Activation("relu"))
model.add(Dense(1024))
model.add(keras.layers.Activation("relu"))
model.add(Dense(1024))
model.add(keras.layers.Activation("relu"))
model.add(Dense(1024))
model.add(keras.layers.Activation("relu"))
model.add(Dense(1024))
model.add(keras.layers.Activation("relu"))
model.add(Dense(n_outputs))
model.add(keras.layers.Activation("sigmoid"))
```

This way you can get the output of the 5th Dense layer before the activation function like this:

```python
first_5_layers_without_last_activation = keras.models.Sequential(model.layers[:2*5])
y_scores = first_5_layers_without_last_activation.predict(X_new)
```

If your model was already trained, that's okay. You can just build a new identical model with the activation layers separated like I just did, and then copy the weights like this:

```python
new_model.set_weights(old_model.get_weights())
```

And then get the output of any layer you want like we just did:

```python
first_5_layers_without_last_activation = keras.models.Sequential(model2.layers[:2*5])
y_scores = first_5_layers_without_last_activation.predict(X_new)
```

Hope this helps.

**Comment by ageron** [comment]
Side note: you probably should use "relu" instead of "sigmoid" in all hidden layers, and 7 dense layers with 1024 neurons each is quite a lot, make sure the model is not overfitting the training set. Lastly, since you are using "binary_crossentropy", you must be doing binary classification, so make sure `n_outputs=1` (unless you really do have multiple binary labels for each input, for example day/night and inside/outside (in which case n_outputs is the number of label pairs, 2 in this example).

**Comment by sarkarachintya** [comment]
Thank you very much Ageron!
Your elaborate explanation will help me a lots!

Best regards,
Achintya

**Comment by ageron** [comment]
I'm glad I could help. :)
